### ISSUE TITLE
Background task is not fired until application restart

### ISSUE THREAD

**Issue by havranek1024** [OrchardCMS/OrchardCore#5954]
It seems that ModularBackgroundService can miss that 'Background Tasks' feature has been activated on a tenant and then it will not fire any background tasks in that tenant until application is restarted.

The ModularBackgroundService will check for changes every minute - if there are new tenants running, or if there were any changes made by BackgroundTaskManager (this is done in GetShellsToUpdate method).

Now, if a tenant was running without 'Background Tasks' feature, it has been saved by ModularBackgroundService in a runningShells collection, and associated with NullChangeToken. This means that it will never be updated by changes in BackgroundTaskManager (because the change token won't match) and the only chance for this tenant to register any background tasks is to be removed from runningShells collection, and added there again in next loop iteration. This may happen, because activating a feature restarts the tenant, but it depends on the timing - if the feature is activated and restart is completed when ModularBackgroundService is sleeping, then it will not notice any change and the background tasks will never be registered.

**Comment by sebastienros** [comment]
@jtkech will answer

**Comment by jtkech** [comment]
Did you experienced it?

We cache the previous running shells and we check

    var released = previousShells.Where(s => s.Released).Select(s => s.Settings.Name).ToArray();

So, normally we update any new running shell and also any previous one that has been released.

So, when a feature is enabled => the shell is released => So normally the related shell will be updated.

Before re-doing some tests, **1st important question, did you already experienced this issue?**

**Comment by jtkech** [comment]
@havranek1024 maybe a misunderstanding

The Bg Tasks feature doesn't register any bg tasks, it just allows to update their settings.

When you enable this feature, the shell that we have cached is marked as released and a **new shell instance** will be built, not the same instance that we have cached which is still marked as released.

Then there are 2 scenarios when we check the shells to update. If the shell is already rebuilt (we only run / update bg tasks on a fully built tenant including its pipeline) it is still seen as released in the cached `previousShells` and will be updated.

 If the shell is not yet rebuilt it is seen as released and will not be updated at this point. But in the next loop it will be seen as a new running shell and then added to the `tenantsToUpdate`.

**Comment by havranek1024** [comment]
I did experience something like I described (that's why I started digging into the code), although cannot reproduce it now. Also I looked at the code again and you're right - the `released` collection will catch the previous instance of shell and will remove it's name from `valid` collection, so it will indeed be refreshed. I missed that in my previous analysis.

So it seems that my problem was caused by something else and I raised a false alarm 😳
Sorry for taking the time and thanks for quick response!
